File: Advent of Code 2021/lib/day7.ex
defmodule Day7 do
end

{:ok, file} = File.read("inputs/day7.in")

crabs =
  file
  |> String.split(",", trim: true)
  |> Enum.map(&String.to_integer/1)

a = Enum.min(crabs)
b = Enum.max(crabs)

a..b
|> Enum.map(fn x ->
  crabs
  |> Enum.map(fn c -> abs(x - c) end)
  |> Enum.sum()
end)
|> Enum.min()
|> IO.inspect(label: "Part 1")

a..b
|> Enum.map(fn x ->
  crabs
  |> Enum.map(fn c ->
    d = abs(x - c)
    d * (d + 1) / 2
  end)
  |> Enum.sum()
end)
|> Enum.min()
|> IO.inspect(label: "Part 2")

File: Advent of Code 2015/day7.py
def f(s, d):
    try:
        return int(s)
    except:
        pass
    op = d[s]
    try:
        return int(op)
    except:
        pass
    if "OR" in op:
        # |
        a, o, b = op.split()
        t = f(a, d) | f(b, d)
    elif "AND" in op:
        # &
        a, o, b = op.split()
        t = f(a, d) & f(b, d)
    elif "NOT" in op:
        # ^ 65535
        o, a = op.split()
        t = f(a, d) ^ 65535
    elif "LSHIFT" in op:
        # <<
        a, o, b = op.split()
        t = f(a, d) << f(b, d)
    elif "RSHIFT" in op:
        # >>
        a, o, b = op.split()
        t = f(a, d) >> f(b, d)
    else:
        t = f(op, d)
    d[s] = t
    return t


D = {}
for line in open("inputs/day7.in"):
    b, a = line.strip().split(" -> ")
    D[a] = b

sa = f("a", D)
print("Part 1:", sa)

e = {}
for line in open("inputs/day7.in"):
    b, a = line.strip().split(" -> ")
    e[a] = b
e["b"] = sa

print("Part 2:", f("a", e))
